### PR TITLE
NETOBSERV-1478: Add eBPF Agent flows filtering capability

### PR DESCRIPTION
**Pull request by msherif1234**
## Description

Add ability to configure flow filtering rule for ebpf agent
## Dependencies

https://github.com/netobserv/netobserv-ebpf-agent/pull/307

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [X] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [X] Does this PR require product documentation?
  * [X] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [X] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [X] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by msherif1234** [comment]
config used for testing
```shell
flowFilter:
        config:
          cidr: 0.0.0.0/0
          destPorts: 0
          protocol: UDP
          sourcePorts: 53
        enable: true
```
image

image

File: apis/flowcollector/v1beta1/flowcollector_types.go [hunk 1]
```diff
@@ -166,6 +167,66 @@ type EBPFMetrics struct {
 	Enable *bool `json:"enable,omitempty"`
 }
 
+// FlowFilterConfig defines the configuration of the flow filter.
+type FlowFilterConfig struct {
+	// FlowFilterIPCIDR defines the IP CIDR to filter flows by.
+	// Example: 10.10.10.0/24 or 100:100:100:100::/64
+	FlowFilterIPCIDR string `json:"cidr,omitempty"`
```
Comment (jotak) [inline apis/flowcollector/v1beta1/flowcollector_types.go:174]: nit: could you rename all the fields here without the `FlowFilter` prefix, so the go names reflect the json names?

File: apis/flowcollector/v1beta1/flowcollector_types.go [hunk 2]
```diff
@@ -166,6 +167,66 @@ type EBPFMetrics struct {
...
+	// +optional
+	FlowFilterICMPType *int `json:"icmpType,omitempty"`
+
+	// FlowFilterICMPCode defines the ICMP code to filter flows by.
+	// +optional
+	FlowFilterICMPCode *int `json:"icmpCode,omitempty"`
+}
+
+// `EBPFFlowFilter` defines the desired eBPF agent configuration regarding flow filtering
+type EBPFFlowFilter struct {
```
Comment (jotak) [inline apis/flowcollector/v1beta1/flowcollector_types.go:222]: This and `FlowFilterConfig` could be merged into a single flat struct

File: apis/flowcollector/v1beta2/flowcollector_types.go [hunk 3]
```diff
@@ -173,6 +174,66 @@ type EBPFMetrics struct {
 	Enable *bool `json:"enable,omitempty"`
 }
 
+// FlowFilterConfig defines the configuration of the flow filter.
+type FlowFilterConfig struct {
+	// FlowFilterIPCIDR defines the IP CIDR to filter flows by.
+	// Example: 10.10.10.0/24 or 100:100:100:100::/64
+	FlowFilterIPCIDR string `json:"cidr,omitempty"`
```
Comment (jotak) [inline apis/flowcollector/v1beta2/flowcollector_types.go:181]: (same comments as for v1beta1)

**Comment by jotak** [comment]
some cosmetic comments other than that lgtm

File: config/samples/flows_v1beta2_flowcollector.yaml [hunk 4]
```diff
@@ -22,6 +22,13 @@ spec:
       interfaces: []
       excludeInterfaces: ["lo"]
       kafkaBatchSize: 1048576
+      #flowFilter:
+      #  config:
+      #    action: Accept
+      #    cidr: 0.0.0.0/0
+      #    protocol: UDP
+      #    sourcePorts: 53
+      #  enable: true
```
Comment (jpinsonneau) [inline config/samples/flows_v1beta2_flowcollector.yaml:31]: That's awesome ! 

Do you think we should write some doc or maybe a blog to highlight real use cases ?

**Review by jpinsonneau — APPROVED** [review]

File: config/samples/flows_v1beta2_flowcollector.yaml (hunk 4 above)
Comment (msherif1234) [reply config/samples/flows_v1beta2_flowcollector.yaml:31]: I did add a doc in ebpf agent with usecases that can become blog if we need to

**Review by jotak — APPROVED** [review]

**Review by jpinsonneau — APPROVED** [review]

**Review by jotak — APPROVED** [review]

**Pull request merged**